### PR TITLE
[SPARK-10620][WIP] Migrate TaskMetrics to accumulators

### PR DESCRIPTION
**Pull request by andrewor14**
There exist two mechanisms to pass metrics from executors to drivers: accumulators and `TaskMetrics`. Currently we send both things to the driver using two separate code paths. This is an unnecessary maintenance burden and makes the code more difficult to follow.

**This patch proposes that we send only accumulator updates to the driver.** Additionally, it reimplements `TaskMetrics` using accumulators such that the new `TaskMetrics` serves mainly as a syntactic sugar to increment and access the values of the underlying accumulators. It migrates the rest of the metrics to adopt the code path already used by the existing `PEAK_EXECUTION_MEMORY`.

While an effort has been made to preserve as much of the public API as possible, there were a few known breaking `@DeveloperApi` changes that would be very awkward to maintain. These are:
- `TaskMetrics#hostname` field was removed; event log was the only consumer
- `ExceptionFailure#taskMetrics` field was replaced with `accumUpdates`
- `SparkListenerExecutorMetricsUpdate#taskMetrics` field was replaced with `accumUpdates`
- `AccumulableInfo#update` changed from `Option[String]` to `Option[Any]`
- `AccumulableInfo#value` changed from `String` to `Option[Any]`

The following event log elements are changed:
- `AccumulableInfo` "Update" field changed from `Option[String]` to `Option[Any]`
- `AccumulableInfo` "Value" field changed from `String` to `Option[Any]`

This is WIP because I would like to add tests for some of the intricate cases that I ran into while implementing this.

---

I broke this down into several smaller PRs:
- #10810 - move classes to their own files for readability
- #10811 - rename 3 fields in `ShuffleWriteMetrics`
- #10815 - refactor TaskMetrics to make it not use public var's
- #10835 - implement TaskMetrics using accumulators
- #10857 - send only accumulator updates to driver, not TaskMetrics

### CODE REVIEW

**Comment by andrewor14** [comment]
retest this please

File: project/MimaExcludes.scala [hunk 1]
```diff
@@ -147,6 +147,16 @@ object MimaExcludes {
...
         ProblemFilters.exclude[MissingMethodProblem]("org.apache.spark.streaming.flume.sink.TransactionProcessor.org$apache$spark$streaming$flume$sink$Logging$$log__=")
+      ) ++ Seq(
+        // SPARK-10620 Migrate TaskMetrics to accumulators
+        ProblemFilters.exclude[MissingMethodProblem](
+          "org.apache.spark.TaskContext.internalMetricsToAccumulators"),
+        ProblemFilters.exclude[MissingMethodProblem](
+          "org.apache.spark.TaskContext.collectInternalAccumulators"),
+        ProblemFilters.exclude[MissingMethodProblem](
+          "org.apache.spark.TaskContext.collectAccumulators"),
+        ProblemFilters.exclude[MissingMethodProblem]("org.apache.spark.Accumulable.this"),
```
Comment (JoshRosen) [inline project/MimaExcludes.scala]: Do you remember which change triggered this error?

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
retest this please

**Comment by nongli** [comment]
This patch is really difficult to review since it combines simple refactoring and more subtle logic. Is it possible to break this up? If there were patches that did simple clean up, those can get reviewed and merged very quickly.

**Comment by andrewor14** [comment]
Yeah it's still WIP but once it's in a more ready state I'll try to break it up.

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
OK, I broke this down into smaller PRs so far:

#10810 - move classes to their own files for readability
#10811 - rename 3 fields in ShuffleWriteMetrics
#10815 - refactor TaskMetrics to make it not use public var's

More coming soon, I'll keep this comment updated.

**Comment by andrewor14** [comment]
By the end of all the smaller patches there will be too much conflict to resolve, so I'm closing this.

**Comment by andrewor14** [comment]
Note to self: DO NOT DELETE THIS BRANCH!!

**Pull request closed without merging**